### DISCUSSION TITLE
Multiapp transfer

### DISCUSSION THREAD

**Discussion by karthichockalingam** [idaholab/moose#24880] [Q&A General]
Hi,

I have a master which has two multi-apps (sub1 and sub2) solving time-dependent problems.
Say you u = 0 in the master, which is passed to sub1, which returns u_new to the mater.
Now, when I pass u from the mater to sub2, it still pass u = 0 (and not u_new). Only in the 
next time, does the master pass u_new to the sub app. So my simulation is lagging behind
by one time step.

How can I fix this?

Thank you.
Karthik.

**Comment by GiudGiud** [comment]
Hello

you could execute the multiapp on time step end to begin with. This will make sure the field sent to the multiapp is updated. 

If you want to synchronize the two applications more, you can turn on fixed point iterations between the parent and child apps in the executioner block. 

You may want to turn verbose_multiapps=true in the executioner block for now to set this up

Guillaume

**Comment by karthichockalingam** [comment]
Hi @GiudGiud 

Thank you for your response. 

I have three multi-apps,

master->app1_sub1
master->app2_sub1->app2_sub2->app2_sub3
master->app3_sub1

Say, 

app1_sub1 solves Equation 1. It needs u3, u2 and provides u1.
app2_sub1(->sub2->sub3) solves Equation 2. It needs u1 and provides u2.
app3_sub1 solves Equation 3. It needs u1, u2 and provides u3.

There are initial conditions for u3 and u1. The master just collects the variables and transfers all the variables to the respective multi-apps. 

Q1) I am not sure, how to order the apps, probably I can execute app1_sub1 on timestep_begin and app2_sub1 on timestep_end. What do I execute app3_sub1 on?

Q2) Perhaps Picard (fixed-point) iteration, as you suggested is the way to go to ensure proper coupling between the variables before moving to the next time step. I am not sure how to set up the convergence criteria. All three variables need to be converged before moving to the next time step. What goes into the executioner block? I am new to using Picard iteration. 

Thank you,
Karthik.

**Reply by GiudGiud** [reply]
Q1)
I would execute app3 on timestep end. There's no way to have u1 updated and at the same time provide an updated u3 to app1
And that's where fixed point iterations (Picard) come in.

btw if the main app does not provide any field of interest to the sub apps you can execute every subapp on timestep end. 

Q2)
There are quite a few parameters to the Picard iterations for multiapps. 
You can see all these parameters, and the several ways to set the convergence criterion (either main app nonlienar residual or a custom postprocessor) in the Executioner parameters under Fixed Point Iterations parameters
https://mooseframework.inl.gov/source/executioners/Steady.html

**Reply by karthichockalingam** [reply]
@GiudGiud Thank you for your response.

The main app only gathers the variables (u1, u2, u3) from the sub-app and does not solve. 
```
[Problem]
  type = FEProblem
  solve = false
[]
```
So only auxiliary variables exist in the main app. However, I would like to set the Picard convergence criterion for u1, u2 and u3 in the main app. Does it make sense to set something like the below in the executioner block of the main app?
```
  picard_max_its = 10
  picard_rel_tol = 1e-6
  picard_abs_tol = 1e-10
```
Kind regards,
Karthi

**Reply by GiudGiud** [reply]
no this wont work.
You ll want to use the `custom_pp` convergence option.
You can create a "change in a variable" postprocessor using the existing ones

**Reply by karthichockalingam** [reply]
@GiudGiud,

I know how to create a post-processor, which I understand returns a single scalar quantity.

Q1) What is the convergence criterion on i.e like an l2 error norm pp of an aux variable? 
Q2) Will the above pp be checked against the previous iteration for convergence?
Q3) Sorry, I am not following what you mean by "change in a variable" pp. Are we changing the aux variable to a regular variable using pp?

 Best,
Karthi

**Reply by GiudGiud** [reply]
use this PP
https://mooseframework.inl.gov/source/postprocessors/ChangeOverTimestepPostprocessor.html

Q1) Yes you should use something like that. Or L1 norm maybe
Q2) You can work to make that happen. Please refer to the fixed point iteration documentation from the Executioner
Q3) No, PPs do not convert from regular to aux or aux to regular.

**Reply by karthichockalingam** [reply]
@GiudGiud 

Thank you. I almost understand how to get it to work. 

Looks like the ChangeOverTimeStepPostprocessor convergence criterion is something like |y^n+1| -  |y^n| provided a specific norm.
Where y is the field variable and n is the Picard iteration number. 

However shouldn't the convergence be on |y^n+1 -  y^n| going to zero instead? 

I think this can't be accomplished via ChangeOverTimePostprocessor as it only takes the difference between scalar qualities and not field variables.

**Reply by GiudGiud** [reply]
oh ok might be another postprocessor. 
You could compute the residuals in the apps with 
https://mooseframework.inl.gov/source/auxkernels/DebugResidualAux.html
then transfer the residuals from the apps to main app then use this to compare them to 0
https://mooseframework.inl.gov/source/postprocessors/ElementL2FunctorError.html

**Reply by karthichockalingam** [reply]
@GiudGiud 

Alternatively,  instead of doing the following

app1_sub1 solves Equation 1. It needs u3, u2 and provides u1.
app2_sub1(->sub2->sub3) solves Equation 2. It needs u1 and provides u2.
app3_sub1 solves Equation 3. It needs u1, u2 and provides u3.

I could do

Master solves u3, provided u1,u2 
(but u3 will be zero for the first iteration as u1=u2=0 but Picard iteration will eventually make u3 to converge)

app1_sub1 solves Equation 1. It needs u3, u2 and provides u1.
app2_sub1(->sub2->sub3) solves Equation 2. It needs u1 and provides u2.

Then I set the Picard convergence criterion directly on u3, which would be a non-linear variable not make use of ChangeOverTimeStepPostprocessor?

**Reply by GiudGiud** [reply]
that works too
